### ISSUE TITLE
How would go about implementing instancing gltf models?

### ISSUE THREAD

**Issue by arcman7** [toji/webgpu-gltf-case-study#4]
I see that you're creating instances at the primitive level, but then how would you go about instancing collections of primitives that make up various gltf models?

**Comment by toji** [comment]
I'm not sure I understand the question? Are you asking how to instance primitives if you have multiple copies of a glTF model in you scene?

**Comment by arcman7** [comment]
Yes that is what I'm asking :) 

I'm new to webGPU so I apologize ahead of time if the question I'm asking doesn't make sense.

From this code:
```
@group(1) @binding(0) var<storage> modelMatrices : array<mat4x4f>;

@vertex
fn vertexMain(@location(0) position : vec3f,
              @builtin(instance_index) instance : u32) -> @builtin(position) vec4f {
  return modelMatrices[instance] * vec4f(position, 1);
}
```

It intuitively makes sense to me how you would go about instancing a gltf model. However, you then go on to later define code that does instancing for nodes that re-use the same primitives, and you're using an index into an array that tracks how many instances there are for a given primitive. From my understanding, you'd some how need access to a second array that contains the world transform matrix at the model level for a given gltf instance. How would you go about grouping primitive instances to a specific gltf instance and making the appropriate transforms to the position vertex in the vertex shader?

**Comment by toji** [comment]
So the key here is that at the instancing level you _don't really care_ about whether the instances come from a single model or two models or 1000 models. All that really matters is that you can identify two primitives are the same and append yet another matrix onto the instance list.

In this end this will have much more to do with how your app handles it's scene graph than the type of code covered in this article. For example: If you have two copies of a mesh in your scene you want to make sure that you're re-using the same buffers/textures/primitives/etc in both cases, but each will likely want it's own copy of the node tree so they can be positioned and posed independently. Also, this probably means you want to move some of the prep work that this article does once on load (like building the instance buffer) and instead make it something that happens either per-frame or that can at least be triggered on each scene graph change. 

Finally you'll want to process all copies of the mesh with the renderer at the same time (effectively treating them as nodes of a larger combined node tree) rather than saying `render(model0); render(model1); render(model2); ...` because you want to have visibility of all the possible duplicate primitives when preparing the scene for drawing.

**Comment by arcman7** [comment]
> All that really matters is that you can identify two primitives are the same and append yet another matrix onto the instance list.

>  If you have two copies of a mesh in your scene you want to make sure that you're re-using the same buffers/textures/primitives/etc in both cases, but each will likely want it's own copy of the node tree so they can be positioned and posed independently.

So if I have M nodes per gltf model, and I have N copies of that model, then I'm going to be creating an array of matrices with size MxN for the instances array of a given primitive?

Edit: Assuming the above is correct - if I wanted to make this operation really fast, would it make sense to set the data of this instance array with a compute shader anytime there's an update to the world state (rotation, scaling and translation)? I'm imagining animating a field of grass in this case.

> Also, this probably means you want to move some of the prep work that this article does once on load (like building the instance buffer) and instead make it something that happens either per-frame or that can at least be triggered on each scene graph change.

I was thinking I'd go with the approach you mentioned earlier where we just have a fixed size buffer that holds the data for the maximum possible number of copies that a given gltf model will use.

**Comment by toji** [comment]
> So if I have M nodes per gltf model, and I have N copies of that model, then I'm going to be creating an array of matrices with size MxN for the instances array of a given primitive?

Yeah, that works for cases where you're just rendering the same model N times. It's tricker if your scene has a variety of models intermixed.

> if I wanted to make this operation really fast, would it make sense to set the data of this instance array with a compute shader anytime there's an update to the world state (rotation, scaling and translation)? I'm imagining animating a field of grass in this case.

For something like grass, given the volume of meshes that you'll be dealing with and the specialized logic that you'll likely want to drive it (with things like fading out at a given distance) you may want to do instancing with a more lightweight structure, such as having each instance be defined by just a position and rotation. Building those programmatically in a compute shader is a great idea, especially because it's unlikely that you'll be placing each blade of grass by hand.

**Comment by arcman7** [comment]
Thank you for the feedback!

> Yeah, that works for cases where you're just rendering the same model N times. It's tricker if your scene has a variety of models intermixed.

Is there a reason why you wouldn't just do this same method for each new model that needs to be instanced?

**Comment by toji** [comment]
You definitely could! It's a design decision that could go either way depending on your needs. The more you can consolidate the number of buffers/pipelines/etc you need to switch between the higher your peak performance can be, but that makes things more rigid and reduces the ability to dynamically update the scene.

**Comment by arcman7** [comment]
Thank you so much! This was great btw! If I get a demo running, I'll be sure you link it here incase you're interested in seeing the performance.
